plugins/amd-kria/fu-amd-kria-device.c

```diff
@@ -0,0 +1,213 @@
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-amd-kria-device.h"
+#include "fu-amd-kria-som-eeprom.h"
+
+typedef struct {
+	FuVolume *esp;
+	FuDeviceLocker *esp_locker;
+	gchar *eeprom_address;
+} FuAmdKriaDevicePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuAmdKriaDevice, fu_amd_kria_device, FU_TYPE_I2C_DEVICE)
+
+#define GET_PRIVATE(o) (fu_amd_kria_device_get_instance_private(o))
+
+static gboolean
+fu_amd_kria_device_prepare(FuDevice *device,
+			   FuProgress *progress,
+			   FwupdInstallFlags flags,
+			   GError **error)
+{
+	FuAmdKriaDevice *self = FU_AMD_KRIA_DEVICE(device);
+	FuAmdKriaDevicePrivate *priv = GET_PRIVATE(self);
+
+	priv->esp_locker = fu_volume_locker(priv->esp, error);
+	if (priv->esp_locker == NULL)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_amd_kria_device_cleanup(FuDevice *device,
+			   FuProgress *progress,
+			   FwupdInstallFlags flags,
+			   GError **error)
+{
+	FuAmdKriaDevice *self = FU_AMD_KRIA_DEVICE(device);
+	FuAmdKriaDevicePrivate *priv = GET_PRIVATE(self);
+
+	if (!fu_device_locker_close(priv->esp_locker, error))
+		return FALSE;
+	g_clear_object(&priv->esp_locker);
+
+	return TRUE;
+}
+
+static gboolean
+fu_amd_kria_device_write_firmware(FuDevice *device,
+				  FuFirmware *firmware,
+				  FuProgress *progress,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	FuContext *ctx = fu_device_get_context(device);
+	g_autofree gchar *cod_path = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	cod_path = g_build_filename(fu_context_get_esp_location(ctx),
+				    "EFI", "UpdateCapsule", "fwupd.cap", NULL);
+	g_debug("using %s for capsule", cod_path);
+	if (!fu_path_mkdir_parent(cod_path, error))
+		return FALSE;
+	if (!fu_bytes_set_contents(cod_path, fw, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_amd_kria_device_set_quirk_kv(FuDevice *device,
+				const gchar *key,
+				const gchar *value,
+				GError **error)
+{
+	FuAmdKriaDevice *self = FU_AMD_KRIA_DEVICE(device);
+	FuAmdKriaDevicePrivate *priv = GET_PRIVATE(self);
+
+	if (g_strcmp0(key, "AmdKriaEeprom") == 0) {
+		priv->eeprom_address = g_strdup(value);
+		return TRUE;
+	}
+	return FU_DEVICE_CLASS(fu_amd_kria_device_parent_class)
+	    ->set_quirk_kv(device, key, value, error);
+}
+
+static gboolean
+fu_amd_kria_device_probe(FuDevice *device, GError **error)
+{
+	FuAmdKriaDevice *self = FU_AMD_KRIA_DEVICE(device);
+	FuAmdKriaDevicePrivate *priv = GET_PRIVATE(self);
+	GUdevDevice *udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	const gchar *tmp;
+	g_auto(GStrv) of_name = NULL;
+
+	/* FuI2cDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_amd_kria_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	tmp = g_udev_device_get_property(udev_device, "OF_FULLNAME");
+	of_name = fu_strsplit(tmp, strlen(tmp), "@", -1);
+	tmp = of_name[g_strv_length(of_name) - 1];
```

Comment:
I think you need to check tmp for NULL -- and maybe add a comment explaining what it's doing?